src/plugins/intel_cpu/src/nodes/mvn.cpp

```diff
@@ -333,97 +244,617 @@ struct jit_uni_mvn_mean_variance_kernel_f32 : public jit_uni_mvn_mean_variance_k
+
+        Xbyak::Label tail_label;
+        cmp(reg_vector_num, 0);
+        je(tail_label, T_NEAR);
+
+        // unroll_size = vec_num >= unroll_size ? unroll_size : vec_num;
+        Xbyak::Label label_reset_unroll_size_end;
+        cmp(reg_unroll_size, reg_vector_num);
+        jle(label_reset_unroll_size_end, T_NEAR);
+        mov(reg_unroll_size, reg_vector_num);
+        L(label_reset_unroll_size_end);
+
+        // last unroll_size
+        Xbyak::Label label_reset_last_unroll_size;
+        Xbyak::Label label_reset_last_unroll_size_end;
+        Xbyak::Reg64 last_unroll_size = rcx;
+        mov(rax, reg_vector_num);
+        xor_(rdx, rdx);
+        div(reg_unroll_size);  // rdx
+        cmp(rdx, 0);
+        je(label_reset_last_unroll_size, T_NEAR);
+        mov(last_unroll_size, rdx);
+        jmp(label_reset_last_unroll_size_end);
+        L(label_reset_last_unroll_size);
+        {
+            mov(last_unroll_size, reg_unroll_size);
+        }
+        L(label_reset_last_unroll_size_end);
+
+        // size_t unroll_number = div_up(vec_num, unroll_size); --> (vec_num + unroll_size - 1) / unroll_size;
+        mov(rdi, reg_vector_num);
+        add(rdi, reg_unroll_size);
+        sub(rdi, 1);
+        mov(rax, rdi);
+        xor_(rdx, rdx);
+        div(reg_unroll_size);  // result is in rax, that is reg_unroll_num, no mov need.
 
         int ur_base = 4;
-        Xbyak::Reg64 reg_src_aux = reg_stride;
-        Xbyak::Reg64 reg_work_amount_bk = rbx;
-        mov(reg_work_amount_bk, reg_work_amount);
-        for (size_t ur_num = 0; ur_num < unroll_number; ur_num++) {
+        Xbyak::Label label_unroll_num;
+        Xbyak::Label label_unroll_num_end;
+        L(label_unroll_num);
+        {
+            cmp(reg_unroll_num, 0);
+            jle(label_unroll_num_end, T_NEAR);
+
+            Xbyak::Label label_not_last;
+            cmp(reg_unroll_num, 1);
+            jne(label_not_last, T_NEAR);
+            mov(reg_unroll_size, last_unroll_size);
+            L(label_not_last);
+
             // 4-15 for unroll. 4-7 for src, 8-11 for m/v sum, 12-15 for mean
-            int ur_offset_elt = ur_num * unroll_size * vector_step;
-            int ur_offset = ur_offset_elt * sizeof(float);
-            size_t unroll_size_rt = std::min(vec_num - ur_num * unroll_size, unroll_size);
-            size_t elt_num = std::min(jcp_.C - ur_num * unroll_size * vector_step, unroll_size * vector_step);
-            for (size_t ur_size = 0; ur_size < unroll_size_rt; ur_size++) {
-                uni_vpxor(Vmm(ur_base + 4 + ur_size), Vmm(ur_base + 4 + ur_size), Vmm(ur_base + 4 + ur_size));
-            }
-            if (jcp_.normalize_variance) {
-                for (size_t ur_size = 0; ur_size < unroll_size_rt; ur_size++) {
-                    uni_vmovups(Vmm(ur_base + 8 + ur_size), ptr[reg_mean + ur_offset + ur_size * vector_step * sizeof(float)]);
-                }
-            }
+            Xbyak::Label label_init_end;
```

Comment:
For initializing vector registers here, AFAIK we can use 'for' loops to avoid code duplication, which will be unrolled during JIT compilation stage.

Reply:
"for loop" is available only we know the iteration number, which is ok for static. For dynamic, we do not know the iteration number in compilation stage. we need dynamically decide if go to next iteration based on runtime shape.

Reply:
@dmitry-gorokhov In case you have potential concern there are perf penalty with more compare, I had experiment shows that the cmp and jmp have nearly zero impact as all hidden in ILP in mvn code context, they are on different execution port.
Instead we can create static 1/2/3/4 unroll code block and compare only once on the top. But this will increase compilation time and compiled kernel size.

Reply:
> "for loop" is available only we know the iteration number, which is ok for static. For dynamic, we do not know the iteration number in compilation stage. we need dynamically decide if go to next iteration based on runtime shape.

I mean, since the unroll factor is 4 for this code block https://github.com/chenhu-wang/openvino/blob/chenhu/shape_agnostic_jit_kernel_for_MVN/src/plugins/intel_cpu/src/nodes/mvn.cpp#L308-L326, why not just use "for loop" to generator the loop body. It seems can get the same code block you write, and avoid duplication.

Reply:
@xuchen-intel , As I mentioned, runtime reg_unroll_size is not always 4 for dynamic case. We can create loops for each possible values. but again, it increase load-time, and kernel footprint size. Experiment shows kernel footprint size have slight perf negative impact.